scanpipe/pipes/d2d.py

```diff
@@ -1108,3 +1118,92 @@ def _map_thirdparty_npm_packages(package_json, to_resources, project):
 
     package_resources.no_status().update(status=flag.NPM_PACKAGE_LOOKUP)
     return package_resources.count()
+
+
+def match_resources_with_no_java_source(project, logger=None):
+    """
+    Match resources with ``no-java-source`` to PurlDB, if no match
+    is found update status to ``requires-review``.
+    """
+    project_files = project.codebaseresources.files()
+
+    to_no_java_source = project_files.to_codebase().filter(status=flag.NO_JAVA_SOURCE)
+
+    if to_no_java_source:
+        resource_count = to_no_java_source.count()
+        if logger:
+            logger(
+                f"Mapping {resource_count:,d} to/ resources with {flag.NO_JAVA_SOURCE} "
+                "status in PurlDB using SHA1"
+            )
+
+        _match_purldb_resources(
+            project=project,
+            to_resources=to_no_java_source,
+            matcher_func=match_purldb_resource,
+            logger=logger,
+        )
+        to_no_java_source.no_status().update(status=flag.REQUIRES_REVIEW)
+
+
+def match_unmapped_resources(project, matched_extensions=None, logger=None):
+    """
+    Match resources with ``too-many-maps`` and empty status to PurlDB,
+    flag resources with empty status as ``requires-review``.
+    """
+    project_files = project.codebaseresources.files()
+
+    to_unmapped = (
+        project_files.to_codebase()
+        .filter(status__in=["", flag.TOO_MANY_MAPS])
+        .exclude(is_media=True)
+    )
+
+    if matched_extensions:
+        to_unmapped.exclude(extension__in=matched_extensions)
+
+    if to_unmapped:
+        resource_count = to_unmapped.count()
+        if logger:
+            logger(
+                f"Mapping {resource_count:,d} to/ resources with {flag.TOO_MANY_MAPS} "
+                "or empty status in PurlDB using SHA1"
+            )
+
+        _match_purldb_resources(
+            project=project,
+            to_resources=to_unmapped,
+            matcher_func=match_purldb_resource,
+            logger=logger,
+        )
+        to_unmapped.no_status().update(status=flag.REQUIRES_REVIEW)
+
+    to_without_status = project_files.to_codebase().no_status()
+
+    to_without_status.filter(is_media=True).update(status=flag.IGNORED_MEDIA_FILE)
+
+    to_without_status.no_status().update(status=flag.REQUIRES_REVIEW)
```

Comment:
It seems that the `.no_status()` is already applied to the initial `to_without_status` var.
Is it actually needed in this last line to trigger the proper update?

Reply:
No, not needed.